### DISCUSSION TITLE
Lo! Phase 2 of a WEB→HTML translator.

### DISCUSSION THREAD

**Discussion by texdraft** [shreevatsa/webWEB#13] [Show and tell]
I have now implemented Phase 2 of my translator. The repository at https://github.com/texdraft/web-to-html has been updated. It's been a while since my last post; simply put, Phase 2 turned out to be a lot trickier to write than I had expected.

Caveats:

- There are almost certainly problems with the line-breaking and indentation logic; these will be uncovered once Phase 3 is written. (The “hard part” is the semantic analysis, which seems to be working.)
- Changes to most of Knuth's `WEB` programs are necessary. (“compiler directives” must be removed, among others).
- It is written in a more imperative style than I would like.
- Error recovery is nonexistent—the program just dies if a syntactic or semantic problem is encountered.
- Inline code is left alone. I still don't know the best strategy for handling it; ideas are welcome.
- Almost all of the Pascal language is supported (including `with` statements and forward references in pointer types); the one exception is that routine parameters aren't allowed. I left them out only because I don't know how to fit them nicely into my heirarchy of `meaning` objects (see `pascal.lisp`).

Here is an excerpt of the code (`if` has already been scanned when this procedure is called).

```lisp
;; <if statement> ::= if <expression> then <statement>
;;                  | if <expression> then <statement> else <statement
(defun parse-if-statement (parser)
  "Parse an if statement."
  (parse-expression parser)
  (let ((token (get-next parser)))
    (unless (token-matches-p token "then")
      (expected-error parser "``then''"))
    (attach-line-break token :indent :after)
    (parse-statement parser)
    (setf token (get-next parser))
    (attach-line-break token :dedent :before)
    (when (token-matches-p token "else")
      (attach-line-break token :indent :after)
      (parse-statement parser)
      (setf token (get-next parser))
      (attach-line-break token :dedent :before))
    (put-back-token parser token)))
```

The complexity of macro and module expansion is hidden behind the `get-next` function.

----

I found the following programs to be helpful while writing the code:

- [Knuth's `PROFILE`](http://mirrors.ctan.org/obsolete/systems/knuth/pxl/profile.web), an obscure `WEB` tool that has a similar task (statically analyzing a Pascal program and pretty-printing it)
- [The P4 Pascal compiler](https://homepages.cwi.nl/~steven/pascal/)

**Comment by shreevatsa** [comment]
Hey this is great progress, thanks for the update and sorry I haven't been able to reply earlier (busy week at work and was planning to reply on the weekend but lost the weekend to something else…). I look forward to completion of Phase 3 so we can try the entire system end-to-end. :-)
